File: solutions/ruby/connect/1/connect.rb
# frozen_string_literal: true

# class Board
class Board
  def initialize(rows) = (@rows = rows.map { _1.delete(' ') }
                          @height = @rows.size
                          @width = @rows.first&.size.to_i)

  def winner
    return 'X' if connected?('X', @height.times.map { [_1, 0] }) { _2 == @width - 1 }
    return 'O' if connected?('O', @width.times.map { [0, _1] }) { _1 == @height - 1 }

    ''
  end

  private

  NEIGHBORS = [[-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0]].freeze
  def connected?(stone, starts)
    stack = starts.select { |r, c| @rows[r]&.[](c) == stone }
    seen = stack.to_h { [_1, true] }

    until stack.empty?
      r, c = stack.pop
      return true if yield(r, c)

      NEIGHBORS.each do |dr, dc|
        nr = r + dr
        nc = c + dc
        next unless nr.between?(0, @height - 1) && nc.between?(0, @width - 1)
        next unless @rows[nr][nc] == stone

        pos = [nr, nc]
        next if seen[pos]

        seen[pos] = true
        stack << pos
      end
    end

    false
  end
end

File: solutions/clojure/connect/1/src/connect.clj
(ns connect)

(def ^:private neighbor-offsets
  [[-1 0] [-1 1] [0 -1] [0 1] [1 -1] [1 0]])

(defn- connected? [board player starts destination?]
  (loop [frontier (vec starts)
         visited #{}]
    (if-let [[row column :as position] (peek frontier)]
      (cond
        (visited position)
        (recur (pop frontier) visited)

        (destination? position)
        true

        :else
        (let [neighbors (for [[row-offset column-offset] neighbor-offsets
                              :let [neighbor [(+ row row-offset)
                                              (+ column column-offset)]]
                              :when (= player (get-in board neighbor))]
                          neighbor)]
          (recur (into (pop frontier) neighbors)
                 (conj visited position))))
      false)))

(defn connect-winner [board]
  (let [board (mapv #(vec (remove #{\space} %)) board)
        height (count board)
        width (count (first board))
        x-starts (for [row (range height)
                       :when (= \X (get-in board [row 0]))]
                   [row 0])
        o-starts (for [column (range width)
                       :when (= \O (get-in board [0 column]))]
                   [0 column])]
    (cond
      (connected? board \X x-starts #(= (second %) (dec width))) :X
      (connected? board \O o-starts #(= (first %) (dec height))) :O
      :else :no-winner)))
